nixos/seafile: add optional seafdav service

Pacman99 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
add webdav support for seafile

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Configuration Example
```nix
{
  services.seafile.seafdavSettings = {
    enabled = true;
    share_name = "/seafdav";
  };
  services.nginx.virtualHosts."files.example.com" = {
    locations."/seafdav".proxyPass = "http://127.0.0.1:6001/seafdav";
  };
}
```

quantmint commented:
Result of `nixpkgs-review pr 152069` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python38Packages.seafdav (python39Packages.seafdav)</li>
    <li>python38Packages.seafobj (python39Packages.seafobj)</li>
  </ul>
</details>

greizgh reviewed (commented):
Thank you for contributing this feature, I'll test and report back.

Pacman99 commented on pkgs/development/python-modules/seafobj/setuptools.patch:
```diff
@@ -0,0 +1,24 @@
++
++setup(name='seafobj',
++      version=__version__,
++      license='BSD',
```
I don't love this method to add the setup.py, because it means I'd have to maintain the patch too. Is there a better way to package a python library without a setuppy?

dotlambda commented on pkgs/development/python-modules/seafdav/default.nix:
```diff
@@ -0,0 +1,41 @@
+{ lib, fetchFromGitHub, seafobj, python3Packages, makeWrapper }:
```
```suggestion
{ lib, fetchFromGitHub, seafobj, python3, makeWrapper }:
```

dotlambda commented on pkgs/development/python-modules/seafdav/default.nix:
```diff
@@ -0,0 +1,41 @@
+{ lib, fetchFromGitHub, seafobj, python3Packages, makeWrapper }:
+
+python3Packages.buildPythonPackage rec {
```
```suggestion
python3.pkgs.buildPythonApplication rec {
```

dotlambda reviewed (changes requested):
`seafdav` shouldn't be in `pythonPackages`, so put it in `all-packages.nix` and move the expression to an appropriate directory

greizgh commented on nixos/tests/seafile.nix:
```diff
@@ -117,5 +123,8 @@ import ./make-test-python.nix ({ pkgs, ... }:
           client2.succeed('[ `cat test01/first_file` = "bla" ]')
+
+      with subtest("seafdav query"):
+          client1.succeed("curl --anyauth --user 'admin@myrdd.info:seafile_password' http://server/seafdav | grep 'Authenticated user' >&2")
```
This should match declared user:
```suggestion
          client1.succeed("curl --anyauth --user 'admin@example.com:seafile_password' http://server/seafdav | grep 'Authenticated user' >&2")
```

Pacman99 committed:
python3Packages.seafobj: init at unstable-2021-07-08

Pacman99 committed:
seafdav: init at unstable-2021-05-09

Pacman99 committed:
nixos/seafile: add optional seafdav service

Pacman99 commented:
Thanks! I addressed most of the code review and fixed the commit messages.

Pacman99 commented:
I'm going to mark as draft until #153388 is merged

dotlambda commented on pkgs/development/python-modules/seafobj/default.nix:
```diff
@@ -0,0 +1,28 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/haiwen/seafobj";
+    description = "python library for accessing seafile data model";
```
```suggestion
    description = "Python library for accessing seafile data model";
```
https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes

dotlambda commented on pkgs/servers/seafdav/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+  meta = with lib; {
+    homepage = "https://github.com/haiwen/seafdav";
+    description = "the WebDAV server for seafile";
```
```suggestion
    description = "WebDAV server for seafile";
```

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -8867,6 +8867,8 @@ in {
     python3 = self.python;
   });
 
+  seafobj = callPackage ../development/python-modules/seafobj { };
```
This file should be alphabetically ordered.

dotlambda commented on pkgs/development/python-modules/seafobj/default.nix:
```diff
@@ -0,0 +1,28 @@
+  ];
+  doCheck = false; # disabled because it requires a ccnet environment
+
+  nativeBuildInputs = [ makeWrapper ];
```
```suggestion
```

dotlambda commented on pkgs/servers/seafdav/default.nix:
```diff
@@ -0,0 +1,41 @@
+{ lib, fetchFromGitHub, python3, makeWrapper }:
```
```suggestion
{ lib, fetchFromGitHub, python3 }:
```

dotlambda commented on pkgs/servers/seafdav/default.nix:
```diff
@@ -0,0 +1,41 @@
+
+  doCheck = false; # disabled because it requires a ccnet environment
+
+  nativeBuildInputs = [ makeWrapper ];
```
```suggestion
```

Pacman99 commented:
Thank you for the review! I applied everything and rebased.

dotlambda commented on pkgs/development/python-modules/seafobj/default.nix:
```diff
@@ -0,0 +1,27 @@
+{ lib, fetchpatch, fetchFromGitHub, python3Packages }:
```
```suggestion
{ lib
, fetchFromGitHub
, buildPythonPackage
}:
```

dotlambda replied:
No `python3Packages`!!
Otherwise python39.pkgs.sefobj and python310.pkgs.seafobj would be the same.

dotlambda commented on pkgs/development/python-modules/seafobj/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+  patches = [
+    ./setuptools.patch
+  ];
+  doCheck = false; # disabled because it requires a ccnet environment
+
+
+  meta = with lib; {
```
```suggestion
  ];

  doCheck = false; # disabled because it requires a ccnet environment

  meta = with lib; {
```

dotlambda reviewed (changes requested):
The last commit message should start with `nixos/seafile`.

Pacman99 commented on pkgs/development/python-modules/seafobj/setuptools.patch:
```diff
@@ -0,0 +1,24 @@
+@@ -0,0 +1,18 @@
++from setuptools import setup, find_packages
++
++__version__ = '7.0.0'
```
I guess I packaged the pre-release of the next version, but then I'd have to predict what the next version number for seafobj would be.

Pacman99 replied:
Filed an issue upstream for adding setuptools support: https://github.com/haiwen/seafobj/issues/57

dotlambda commented on pkgs/development/python-modules/seafobj/default.nix:
```diff
@@ -0,0 +1,31 @@
+{ lib
+, fetchpatch
```
```suggestion
```
as requested before

Pacman99 replied:
Whoops must have missed that

dotlambda commented:
@Pacman99 Is this done? You should sqash the last commit into the appropriate one and as said above, the third commit message should start with `nixos/seafile`.

Pacman99 commented:
> @Pacman99 Is this done? You should sqash the last commit into the appropriate one and as said above, the third commit message should start with `nixos/seafile`.

I think so, I think I can deal with the setuptools issue in the future, hopefully with an upstream fix. And I believe I've addressed all your review, let me know if I'm missing anything else.

shimunn commented:
What's blocking this, any help needed?
